In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("HW0.ipynb")

# EECS C106A Python Bootcamp
### EECS C106A: Introduction to Robotics, Fall 2026

# Introduction

We will be using the Python programming language for lab and homework assignments in EECS/BioE/MechE C106A/206A. Some homework assignments will entail matrix calculations where Python will come in handy, but you are welcome to use something like MATLAB instead. This notebook is meant to be a mini bootcamp on Python for students who have had experience programming in another language already (e.g. MATLAB) or need a quick refresher on Python. We will use NumPy for numerical calculations. If you have experience with MATLAB but not Python, see [NumPy for MATLAB users](https://numpy.org/doc/stable/user/numpy-for-matlab-users.html).

## Before you begin

Follow the Python 3.13 environment instructions in the repository `README.md`, confirm that `otter` and `numpy` import successfully, and launch JupyterLab from the repository root. Then open `hw0/HW0.ipynb`.

# Table of Contents
* [IPython Basics](#IPython-Basics)
* [Python Data Types](#Data-Types)
* [Python Functions](#Functions)

# IPython Basics

For those who have never used IPython, it is a command shell for interactive Python programming. You can imagine it as allowing you to run blocks of Python code that you would execute in a single Python script (`python script_name.py`) in the terminal. Benefits of using IPython include easier visualization and debugging. The purpose of this bootcamp in IPython is to give you an idea of basic Python syntax and semantics. For labs you are still expected to write and execute actual Python scripts to work with ROS.

### Executing Cells

Jupyter Notebooks are made of cells containing text, code, or HTML. To run a cell, click it and press Shift+Enter or use the run button on the toolbar. Run the cell below; you should expect an output of 6:

In [2]:
1+2+3

6

### Import a library

To import a certain library `x`, just type `import x`
Calling function `y` from that library is simply `x.y`
To give the library a different name (e.g. abbreviation), type `import x as z`

In [3]:
import numpy as np
np.add(3, 4)


np.int64(7)

# Python

## Data Types

### Integers and Floats

In Python 2, integer division returns the floor. In Python 3, `/` performs true division and `//` performs floor division. You can read more in the [Python 3 documentation](https://docs.python.org/3/tutorial/introduction.html#numbers).

In [4]:
5 / 4

1.25

In [5]:
5.0 / 4

1.25

### Booleans

Python implements all usual operators for Boolean logic. English, though, is used rather than symbols (no &, ||, etc.). Pay attention to the following syntax, try to guess what the output for each print statement should be before running the cell.

In [6]:
print(0 == False)

t = True
print(1 == t)

print(0 != t)

print(t != 1)

if t is True:
    print(0 != 0)

True
True
True
False
False


### Strings

Strings are supported very well. To concatenate strings we can do the following:

In [7]:
hello = 'hello'
robot = 'robot'

print(hello + ' ' + robot + str(1))

hello robot1


To find the length of a string use `len(...)`

In [8]:
print(len(hello + robot))

10


## Functions

Python functions are defined using the `def` keyword. For example:

In [9]:
def hello_robot(robot_name, yell=True):
    if yell:
        print('HELLO, %s!' % robot_name.upper())
    else:
        print('hello, %s' % robot_name)

In [10]:
hello_robot('Baxter') # Prints "HELLO, BAXTER!"
hello_robot('Sawyer', yell=False)  # Prints "hello, Sawyer"

HELLO, BAXTER!
hello, Sawyer


### Lists

A list is a mutable array of data, meaning we can alter it after instantiating it. To create a list, use square brackets `[]` and fill it with elements.

Key operations:
- `'+'` appends lists
- `len(y)` to get length of list y
- `y[0]` to index into 1st element of y **Python indices start from 0
- `y[1:6]` to slice elements 1 through 5 of y

In [11]:
y = ["Robots are c"] + [0, 0, 1]
y

['Robots are c', 0, 0, 1]

In [12]:
len(y)

4

In [13]:
y[0]

'Robots are c'

In [14]:
def element_slicing(lst):
    # TODO: slice the first three elements of the input list 'lst' 
    # and store in a new list called 'lst_sliced', then print the 
    # 2nd element of this new list  
    lst_sliced = lst[0:3]
    print(lst_sliced[1])

element_slicing(y) # This should print '0'


0


In [15]:
grader.check("element_slicing")

element_slicing results: All test cases passed!

### Loops

You can loop over the elements of a list like this:

In [16]:
robots = ['baxter', 'sawyer', 'turtlebot']
for robot in robots:
    print(robot)
# Prints "baxter", "sawyer", "turtlebot", each on its own line.

baxter
sawyer
turtlebot


If you want access to the index of each element within the body of a loop, use the built-in [`enumerate`](https://docs.python.org/3/library/functions.html#enumerate) function:

In [17]:
robots = ['baxter', 'sawyer', 'turtlebot']

def display_enumerate(lst):
    # TODO: Using a for loop and the python built-in enumerate function,
    # print the enumerated representation of the list
    # e.g. "#1: baxter", "#2: sawyer", "#3: turtlebot", 
    # each on its own line
    for index, item in enumerate(lst, start=1):
        print(f"#{index}: {item}")
    
display_enumerate(robots)


#1: baxter
#2: sawyer
#3: turtlebot


In [18]:
grader.check("display_enumerate")

display_enumerate results: All test cases passed!

### NumPy Array

A NumPy array is like a list with multidimensional support and additional operations described in the [NumPy reference](https://numpy.org/doc/stable/reference/index.html).

NumPy arrays can be manipulated with many arithmetic operations. You can think of them as more powerful lists. Many functions that work with lists also extend to NumPy arrays.

To use NumPy functions, import the package with `import numpy`. By convention, we use the shorter name `np`, as shown above.

### Making a NumPy Array

In [19]:
x = np.array([[1, 2, 3], [4 , 5, 6], [7, 8, 9]])
print(x)
# x is a 3x3 matrix here

[[1 2 3]
 [4 5 6]
 [7 8 9]]


### Finding the shape of a NumPy Array

In [20]:
x.shape # returns the dimensions of the NumPy array

(3, 3)

### Building arrays by filling in entries

Instead of writing out every entry by hand, its sometimes easier to start from an array of all zeroes or the identity matrix, then fill in the only entries you care about.

In [21]:
# Start with a 3x3 matrix of all zeros
rot = np.zeros((3, 3))
theta = np.pi / 4

# Fill in just the entries we need
rot[0, 0] = np.cos(theta)
rot[0, 1] = -np.sin(theta)
rot[1, 0] = np.sin(theta)
rot[1, 1] = np.cos(theta)
rot[2, 2] = 1

print(rot)

# np.eye(n) gives you an n x n identity matrix to start from
# instead of zeros, which is handy when most of the diagonal
# should stay 1
g = np.eye(4)
g[0, 3] = 5   # set the x-position
g[1, 3] = 2   # set the y-position
print(g)

[[ 0.70710678 -0.70710678  0.        ]
 [ 0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]
[[1. 0. 0. 5.]
 [0. 1. 0. 2.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


### Elementwise operations

Arithmetic operations on NumPy arrays correspond to elementwise operations.

In [22]:
print(x)
print
print(x * 5) # NumPy applies the operation to all elements.

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[ 5 10 15]
 [20 25 30]
 [35 40 45]]


### Matrix multiplication

In [23]:
print(np.dot(x, x))

[[ 30  36  42]
 [ 66  81  96]
 [102 126 150]]


### Slicing NumPy arrays

NumPy slices are usually views into an existing array rather than implicit copies. This is helpful for large arrays, but modifying a view can also modify the original array. Make an explicit copy when that is not intended.

In [24]:
orig = np.array([0, 1, 2, 3, 4, 5])
print(orig)

[0 1 2 3 4 5]


Slicing an array is just like slicing a list

In [25]:
sliced = orig[1:4]
print(sliced)

[1 2 3]


Note, since slicing does not copy the array, mutating `sliced` mutates `orig`. Notice how the 4th element in `orig` changes to 9 as well.

In [26]:
sliced[2] = 9
print(orig)
print(sliced)

[0 1 2 9 4 5]
[1 2 9]


We should use `np.copy()` to actually copy `orig` if we don't want to mutate it. 

In [27]:
orig = np.array([0, 1, 2, 3, 4, 5])
copy = np.copy(orig)
sliced_copy = copy[1:4]
sliced_copy[2] = 9
print(orig)
print(sliced_copy)

[0 1 2 3 4 5]
[1 2 9]


### Slicing in more than one dimension

Everything above works in 1D arrays, but NumPy slicing also works in multiple dimensions: `arr[row_slice, col_slice]`. You can assign directly into a slice to set a whole block at once. This is useful when building larger matrices from smaller ones.

In [28]:
g = np.eye(4)
rot = np.array([[0, -1, 0], [1, 0, 0], [0, 0, 1]])
position = np.array([1, 2, 3])

# Assign the top-left 3x3 block to be our matrix rot
g[:3, :3] = rot

# Assign the top 3 entries of the last column to be our array position
g[:3, 3] = position

print(g)

# You can read a block back out the same way
print(g[:3, :3]) 
print(g[:3, 3])

[[ 0. -1.  0.  1.]
 [ 1.  0.  0.  2.]
 [ 0.  0.  1.  3.]
 [ 0.  0.  0.  1.]]
[[ 0. -1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]]
[1. 2. 3.]


### Reshaping arrays

.reshape() lets you view the same data but with a different shape, which comes in handy when a function expects a specific shape for its input.

In [29]:
v = np.array([1, 2, 3])
print(v.shape)          # (3,)

v_col = v.reshape(3, 1)
print(v_col.shape)      # (3, 1)
print(v_col)

# reshape doesn't copy data -- it's a view, just like slicing
v_col[0, 0] = 99
print(v)                # v is changed too!

(3,)
(3, 1)
[[1]
 [2]
 [3]]
[99  2  3]


In [30]:
A = np.array([[5, 6, 8], [2, 4, 5], [3, 1, 10]])
B = np.array([[3, 5, 0], [3, 1, 1]])

def matrix_question():
    # TODO: multiply matrix A with matrix B padded with 1's to the 
    # same dimensions as A; sum this result with the identity matrix
    # (you may find np.concatenate, np.vstack, np.hstack, or np.eye useful). 
    # Make sure you don't alter the original contents of B. Print the result
    ones = np.ones(3)
    B_pad = np.vstack((B, ones))
    result = np.matmul(A, B_pad)
    idm = np.eye(3)
    print(result + idm)

matrix_question()


[[42. 39. 14.]
 [23. 20.  9.]
 [22. 26. 12.]]


In [31]:
grader.check("matrix")

matrix results: All test cases passed!

### Handy NumPy function: arange

We use `arange` to create integer sequences as NumPy arrays. It is similar to Python's built-in `range`, but returns a NumPy array rather than a list.

`arange(0,N)` instantiates an array listing every integer from 0 to N-1.

`arange(0,N,i)` instantiates an array listing every `i` th integer from 0 to N-1 .

In [32]:
print(np.arange(-3,4)) # every integer from -3 ... 3

[-3 -2 -1  0  1  2  3]


In [33]:
def arange_question(x):
    # TODO: print twice each even integer from 0 through x (inclusive)
    # as a NumPy array. For example, x=3 should print [0 4].
    print(np.arange(0, x + 1, 2) * 2)

arange_question(6)

[ 0  4  8 12]


In [34]:
grader.check("arange")

arange results: All test cases passed!

# References
- [1] EE 120 lab1
- [2] EECS 126 Lab01
- [3] EE 16a Python Bootcamp
- [4] CS231n Python NumPy Tutorial. [Link](https://cs231n.github.io/python-numpy-tutorial/)

## Submission

1. Restart the kernel and run all cells from top to bottom.
2. Confirm that the final check reports that all available tests pass.
3. Save **hw0/HW0.ipynb**.
4. Commit and push the notebook to your private semester repository.
5. In Gradescope, submit that repository and its `main` branch using the GitHub submission option. Do not create or upload a zip file.

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [35]:
grader.check_all()

arange results: All test cases passed!

display_enumerate results: All test cases passed!

element_slicing results: All test cases passed!

matrix results: All test cases passed!